diff --git i/engine/contrib/init/systemd/docker.service w/engine/contrib/init/systemd/docker.service
index d8c7867057..b73ecf7363 100644
--- i/engine/contrib/init/systemd/docker.service
+++ w/engine/contrib/init/systemd/docker.service
@@ -30,6 +30,10 @@ StartLimitInterval=60s
 # in the kernel. We recommend using cgroups to do container-local accounting.
 LimitNPROC=infinity
 LimitCORE=infinity
+# Older systemd versions default to a LimitNOFILE of 1024:1024, which is insufficient for many
+# applications including dockerd itself and will be inherited. Raise the hard limit, while
+# preserving the soft limit for select(2).
+LimitNOFILE=1024:524288
 
 # Comment TasksMax if your systemd version does not support it.
 # Only systemd 226 and above support this option.
